File: CPP/ByTier/F/2128_F_Strict_Triangle.cpp
// Problem: CF 2128 F - Strict Triangle
// https://codeforces.com/contest/2128/problem/F

/*
F. Strict Triangle
Algorithms/Techniques: Dijkstra's algorithm, graph traversal with priority queues

Time Complexity: O((n + m) * log n) per test case
Space Complexity: O(n + m)

This problem determines whether there exists an assignment of edge weights such that 
the shortest path from node 1 to node n does NOT equal the sum of shortest paths 
from 1 to k and k to n. We use two Dijkstra runs:
1. First to compute shortest distances from k to all nodes (using upper bounds r[i]).
2. Second to compute a "critical" value dp[v] which represents the minimum cost needed 
   to ensure that dist(1,v) > dist(1,k) + dist(k,v).
The key insight is that we want to make the triangle inequality strict, so we check 
if there's any assignment that allows strict inequality.

*/
#include <stdio.h>
#include <algorithm>
#include <functional>
#include <queue>
#include <tuple>
#include <utility>
#include <vector>

#define rep(i, a, b) for (int i = (a); i <= (b); i++)
#define per(i, a, b) for (int i = (a); i >= (b); i--)
#define ll long long
#define sz(x) ((int)x.size())
using namespace std;
using pii = pair<ll, int>;
using tup = tuple<int, int, int>;
const int N = 200005;
int n, m, k, l[N], r[N];
vector<pii> G[N];
void solve() {
  scanf("%d%d%d", &n, &m, &k);
  rep(i, 1, n) G[i].clear();
  rep(i, 1, m) {
    int x, y;
    scanf("%d%d%d%d", &x, &y, &l[i], &r[i]);
    G[x].emplace_back(y, i), G[y].emplace_back(x, i);
  }
  vector<ll> dis(n + 1); // distances from node k to all nodes
  vector<int> vs(n + 1); // visited array
  priority_queue<pii, vector<pii>, greater<pii>> q; // min-heap for Dijkstra
  rep(i, 1, n) dis[i] = 1e18;
  dis[k] = 0;
  q.emplace(0, k);
  while (!q.empty()) {
    int u = q.top().second;
    q.pop();
    if (vs[u])
      continue;
    vs[u] = 1;
    for (auto v : G[u])
      if (dis[v.first] > dis[u] + r[v.second]) {
        dis[v.first] = dis[u] + r[v.second];
        q.emplace(dis[v.first], v.first);
      }
  }
  rep(i, 1, n) vs[i] = 0;
  vector<ll> dp(n + 1); // critical values for maintaining strict inequality
  rep(i, 2, n) dp[i] = 1e18;
  dp[1] = -1e18;
  q.emplace(-1e18, 1);
  while (!q.empty()) {
    int u = q.top().second;
    q.pop();
    if (vs[u])
      continue;
    vs[u] = 1;
    if (dis[u] <= dp[u])
      continue;
    for (auto v : G[u])
      if (v.first != k) { // avoid going back to k to prevent cycles
        ll nv = max(dp[u] + l[v.second], l[v.second] - dis[u]);
        if (nv < dp[v.first])
          dp[v.first] = nv, q.emplace(dp[v.first], v.first);
      }
  }
  if (dis[n] > dp[n])
    puts("YES");
  else
    puts("NO");
}
int main() {
  int T;
  scanf("%d", &T);
  while (T--)
    solve();
}


// https://github.com/VaHiX/codeForces/

File: Java/ByTier/F/2128_F_Strict_Triangle.java
// Problem: CF 2128 F - Strict Triangle
// https://codeforces.com/contest/2128/problem/F

/**
 * Problem: F. Strict Triangle
 * 
 * Purpose:
 * Given an undirected connected graph with m edges, each edge has a weight range [l_i, r_i].
 * For a given node k, determine if there exists an assignment of weights such that
 * the shortest path from node 1 to node n is strictly not equal to the sum of:
 *   - Shortest path from node 1 to node k
 *   - Shortest path from node k to node n
 * 
 * This problem uses two Dijkstra-like algorithms:
 * 1. First, we compute minimum distances from node 1 using edge weights l_i (lower bounds).
 * 2. Then we try to compute maximum possible "penalty" values on paths, where the penalty is
 *    defined as max(current_penalty + weight_on_edge, -optimal_distance_to_node_v)
 * 
 * Algorithms:
 * - Modified Dijkstra with priority queue for finding shortest paths
 * - Two Dijkstra runs:
 *   1. Find minimal path distances from node 1 (using lower bounds of edge weights)
 *   2. Find maximum valid penalties to make a strict inequality.
 * 
 * Time Complexity: O((n + m) * log n) per test case
 * Space Complexity: O(n + m)
 */
import java.io.*;
import java.util.*;
public class Main {
    static class F {
        final private int A = 1 << 16;
        private InputStream b;
        private byte[] c;
        private int d, e;
        public F() {
            b = System.in;
            c = new byte[A];
            d = e = 0;
        }
        public int a() throws IOException {
            int f = 0;
            byte g = h();
            while (g <= 32) g = h();
            boolean i = g == '-';
            if (i) g = h();
            do f = f * 10 + g - '0';
            while ((g = h()) >= '0' && g <= '9');
            return i ? -f : f;
        }
        public long l() throws IOException {
            long f = 0;
            byte g = h();
            while (g <= 32) g = h();
            boolean i = g == '-';
            if (i) g = h();
            do f = f * 10 + g - '0';
            while ((g = h()) >= '0' && g <= '9');
            return i ? -f : f;
        }
        private void m() throws IOException {
            e = b.read(c, d = 0, A);
            if (e == -1) c[0] = -1;
        }
        private byte h() throws IOException {
            if (d == e) m();
            return c[d++];
        }
        public void z() throws IOException {
            if (b == null) return;
            b.close();
        }
    }
    static class J {
        int a, b, c;
        public J(int d, int e, int f) {
            a = d; b = e; c = f;
        }
    }
    static class K implements Comparable<K> {
        long a;
        int b;
        public K(long x, int y) {
            a = x; b = y;
        }
        public int compareTo(K o) {
            return Long.compare(a, o.a);
        }
    }
    static class L {
        int a, b, c;
        public L(int x, int y, int z) {
            a = x; b = y; c = z;
        }
    }
    static final long X = (long) 1e16;
    public static void main(String[] args) throws IOException {
        F r = new F();
        PrintWriter w = new PrintWriter(new BufferedOutputStream(System.out));
        int t = r.a();
        while (t-- > 0) {
            int n = r.a();
            int m = r.a();
            int k = r.a() - 1;
            ArrayList<J>[] g = new ArrayList[n];
            for (int i = 0; i < n; i++) g[i] = new ArrayList<>();
            for (int i = 0; i < m; i++) {
                int u = r.a() - 1;
                int v = r.a() - 1;
                int a = r.a();
                int b = r.a();
                g[u].add(new J(v, a, b));
                g[v].add(new J(u, a, b));
            }
            long[] d = new long[n];
            Arrays.fill(d, X);
            PriorityQueue<K> p = new PriorityQueue<>();
            d[k] = 0;
            p.add(new K(0, k));
            while (!p.isEmpty()) {
                K q = p.poll();
                if (d[q.b] != q.a) continue;
                for (J j : g[q.b]) {
                    if (d[j.a] > q.a + j.c) {
                        d[j.a] = q.a + j.c;
                        p.add(new K(d[j.a], j.a));
                    }
                }
            }
            long[] y = new long[n];
            Arrays.fill(y, X);
            PriorityQueue<K> z = new PriorityQueue<>();
            y[0] = -d[0];
            z.add(new K(y[0], 0));
            while (!z.isEmpty()) {
                K q = z.poll();
                if (y[q.b] != q.a) continue;
                for (J j : g[q.b]) {
                    long u = Math.max(q.a + j.b, -d[j.a]);
                    if (u >= d[j.a] || y[j.a] <= u) continue;
                    y[j.a] = u;
                    z.add(new K(u, j.a));
                }
            }
            w.println(y[n - 1] != X ? "Yes" : "No");
        }
        w.flush();
        w.close();
        r.z();
    }
}


// https://github.com/VaHiX/CodeForces/